### PR TITLE
Idr solver

### PR DESCRIPTION
**Pull request by fritzgoebel**
This PR adds the IDR(s) solver.

The solver works for general matrices, it takes factory parameters `kappa` (for a "maintaining the convergence" strategy, default 0.7) and `subspace_dim` (the s in the name, determines how many intermediate vectors are computed in each iteration, default is 2)

- [x] Write documentation
- [x] Optimize generation of random matrix on GPUs as at the moment, the generate step takes about an order of magnitude more time then the apply step
- [x] Merge PR #655 
- [x] Merge PR #658 
- [x] Merge hiprand contaiers

### CODE REVIEW

File: include/ginkgo/core/solver/idr.hpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        stop_criterion_factory_ = std::move(other);
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Criterion factories.
+         */
+        std::vector<std::shared_ptr<const stop::CriterionFactory>>
+            GKO_FACTORY_PARAMETER(criteria, nullptr);
```
Comment (yhmtsai) [inline include/ginkgo/core/solver/idr.hpp:136]: We use `GKO_FACTORY_PARAMETER_VECTOR` or `GKO_FACTORY_PARAMETER_SCALAR` in ginkgo. `GKO_FACTORY_PARAMETER` is only left for keeping interface.
```suggestion
            GKO_FACTORY_PARAMETER_VECTOR(criteria, nullptr);
```
The other parameters should be also changed

File: include/ginkgo/core/solver/idr.hpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        /**
+         * Already generated preconditioner. If one is provided, the factory
+         * `preconditioner` will be ignored.
+         */
+        std::shared_ptr<const LinOp> GKO_FACTORY_PARAMETER(
+            generated_preconditioner, nullptr);
+
+        /**
+         * Dimension of the subspace S. Determines how many intermediate
+         * residuals are computed in each iteration.
```
Comment (yhmtsai) [inline include/ginkgo/core/solver/idr.hpp:153]:
```suggestion
         * Dimension of the subspace S. Determine how many intermediate
         * residuals are computed in each iteration.
```

File: include/ginkgo/core/solver/idr.hpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+         */
+        size_type GKO_FACTORY_PARAMETER(subspace_dim, 2u);
+
+        /**
+         * Threshold to determine if Av_n and v_n are too close to being
+         * perpendicular.
```
Comment (yhmtsai) [inline include/ginkgo/core/solver/idr.hpp:159]: Do you have the equation for it?

File: include/ginkgo/core/solver/idr.hpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        } else {
+            set_preconditioner(matrix::Identity<ValueType>::create(
+                this->get_executor(), this->get_size()[0]));
+        }
+        stop_criterion_factory_ =
+            stop::combine(std::move(parameters_.criteria));
+
+        subspace_dim_ = parameters_.subspace_dim;
+
+        auto subspace_vectors_data = matrix_data<ValueType>(
```
Comment (yhmtsai) [inline include/ginkgo/core/solver/idr.hpp:199]: Do you use subspace like col-major?

Comment (yhmtsai) [reply include/ginkgo/core/solver/idr.hpp:199]: Could the subspace input from user?

**Review by yhmtsai — COMMENTED** [review]
the first part of my review.
I did not go to the algorithm detail

**Comment by fritzgoebel** [comment]
The checks on this will still fail as long as hiprand is not installed on the CI system. I also still need to figure out what to do for hip on NVIDIA hardware.

File: common/solver/idr_kernels.hpp.inc [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+template <typename ValueType>
+__global__ __launch_bounds__(config::warp_size) void set_f_zeros(
+    size_type k, size_type nrhs, ValueType *__restrict__ f_values,
+    size_type f_stride)
+{
+    const auto tidx = thread::get_thread_id_flat();
+    if (tidx < nrhs) {
+        f_values[k * f_stride + tidx] = zero<ValueType>();
+    }
+}
```
Comment (yhmtsai) [inline common/solver/idr_kernels.hpp.inc:333]: fill the value in contiguous memory, so it could be replaced by fill_array

File: common/solver/idr_kernels.hpp.inc [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+
+    if (global_id >= nrhs) {
+        return;
+    }
+
+    if (!stop_status[global_id].has_stopped()) {
+        auto thr = omega[global_id];
+        omega[global_id] /= tht[global_id];
+        auto absrho =
+            abs(thr / (sqrt(real(tht[global_id])) * residual_norm[global_id]));
```
Comment (yhmtsai) [inline common/solver/idr_kernels.hpp.inc:354]: does tht always be real? or always ignore the complex part?

File: core/solver/idr.cpp [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    auto omega = Vector::create(exec, gko::dim<2>{1, nrhs});
+    auto residual_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto tht = Vector::create(exec, dim<2>{1, nrhs});
+    auto t_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto alpha = Vector::create(exec, gko::dim<2>{1, nrhs});
+
+    bool one_changed{};
+    Array<stopping_status> stop_status(exec, nrhs);
+
+    auto subspace_vectors_ =
```
Comment (yhmtsai) [inline core/solver/idr.cpp:134]: subspace_vectors is col-major, right? Could you add some description about it because we usually use row-major?

File: core/solver/idr.cpp [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+        if (stop_criterion->update()
+                .num_iterations(total_iter)
+                .residual(residual.get())
+                .solution(dense_x)
+                .check(RelativeStoppingId, true, &stop_status, &one_changed)) {
+            break;
+        }
+
+        subspace_vectors_->apply(residual.get(), f.get());
```
Comment (yhmtsai) [inline core/solver/idr.cpp:177]: Maybe I am wrong, but I thought subspace_vectors only be transposed not conjugated.

**Review by yhmtsai — CHANGES_REQUESTED** [review]

File: cuda/solver/idr_kernels.cu [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+void solve_lower_triangular(const matrix::Dense<ValueType> *m,
+                            const matrix::Dense<ValueType> *f,
+                            matrix::Dense<ValueType> *c,
+                            const Array<stopping_status> *stop_status)
+{
+    const auto subspace_dim = m->get_size()[0];
+    const auto nrhs = m->get_size()[1] / subspace_dim;
+
+    const auto grid_dim = ceildiv(nrhs, default_block_size);
+    solve_lower_triangular_kernel<<<grid_dim, default_block_size>>>(
```
[the remark this answers is not in the document]
Comment (fritzgoebel) [reply cuda/solver/idr_kernels.cu:134]: I guess it could work. I used the same approach as we use for gmres when solving the small upper triangular system (hessenberg) there.

File: common/solver/idr_kernels.hpp.inc [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+    size_type k, size_type num_rows, size_type subspace_dim, size_type nrhs,
+    const ValueType *__restrict__ residual_values, size_type residual_stride,
+    const ValueType *__restrict__ c_values, size_type c_stride,
+    const ValueType *__restrict__ g_values, size_type g_stride,
+    ValueType *__restrict__ v_values, size_type v_stride,
+    const stopping_status *__restrict__ stop_status)
+{
+    const auto global_id = thread::get_thread_id_flat();
+    const auto row = global_id / v_stride;
+    const auto col = global_id % v_stride;
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply common/solver/idr_kernels.hpp.inc:151]: I guess maybe you can test both and see which performs better in nvprof ?

File: common/solver/idr_kernels.hpp.inc (hunk 6 above)
Comment (fritzgoebel) [reply common/solver/idr_kernels.hpp.inc:354]: tht will always be real, but can be stored in a complex value type. It's the same as norm(t) squared, so since the norm is also needed, doing this saves one time computing th * t

File: include/ginkgo/core/solver/idr.hpp [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+                this->get_executor(), this->get_size()[0]));
+        }
+        stop_criterion_factory_ =
+            stop::combine(std::move(parameters_.criteria));
+
+        subspace_dim_ = parameters_.subspace_dim;
+
+        kappa_ = parameters_.kappa;
+
+        deterministic_ = parameters_.deterministic;
```
Comment (pratikvn) [inline include/ginkgo/core/solver/idr.hpp:214]:
```suggestion
        subspace_dim_ = parameters_.subspace_dim;
        kappa_ = parameters_.kappa;
        deterministic_ = parameters_.deterministic;
```
I dont think newlines are needed here.

File: omp/CMakeLists.txt [hunk 12]
```diff
@@ -7,8 +7,8 @@ target_sources(ginkgo_omp
     components/fill_array.cpp
     components/precision_conversion.cpp
     components/prefix_sum.cpp
-    factorization/ilu_kernels.cpp
     factorization/factorization_kernels.cpp
+    factorization/ilu_kernels.cpp
```
Comment (pratikvn) [inline omp/CMakeLists.txt:11]: Why was this added ?

File: omp/test/solver/idr_kernels.cpp [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+
+    Idr() : rand_engine(30) {}
+
+    void SetUp()
+    {
+        ref = gko::ReferenceExecutor::create();
+        omp = gko::OmpExecutor::create();
+
+        mtx = gen_mtx(123, 123);
+        make_diag_dominant(mtx.get());
```
Comment (pratikvn) [inline omp/test/solver/idr_kernels.cpp:72]: Do we need to make the matrix diagonally dominant ?

File: reference/solver/idr_kernels.cpp [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The BiCGSTAB solver namespace.
```
Comment (pratikvn) [inline reference/solver/idr_kernels.cpp:50]:
```suggestion
 * @brief The IDR solver namespace.
```

File: reference/test/solver/idr_kernels.cpp [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+class Idr : public ::testing::Test {
+protected:
+    using value_type = T;
+    using Mtx = gko::matrix::Dense<value_type>;
+    using Solver = gko::solver::Idr<value_type>;
+
+    Idr()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(gko::initialize<Mtx>(
+              {{1.0, -3.0, 0.0}, {-4.0, 1.0, -3.0}, {2.0, -1.0, 2.0}}, exec)),
```
Comment (pratikvn) [inline reference/test/solver/idr_kernels.cpp:65]: I would also recommend to add a simple test to check for solution of non-symmetric systems.

**Review by pratikvn — COMMENTED** [review]
First part of the review. Only some minor things otherwise looks good!

File: core/solver/idr.cpp (hunk 8 above)
Comment (fritzgoebel) [reply core/solver/idr.cpp:177]: True, but as they are randomly generated, I decided to just generate P^H and skip the conjugating.

File: cuda/base/curand_bindings.hpp [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+            gen, reinterpret_cast<remove_complex<ValueType> *>(values), n,   \
+            mean, stddev));                                                  \
+    }                                                                        \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+
+GKO_BIND_CURAND_RANDOM_VECTOR(float, curandGenerateNormal);
+GKO_BIND_CURAND_RANDOM_VECTOR(double, curandGenerateNormalDouble);
+GKO_BIND_CURAND_RANDOM_VECTOR(std::complex<float>, curandGenerateNormal);
```
[the remark this answers is not in the document]
Comment (fritzgoebel) [reply cuda/base/curand_bindings.hpp:99]: No, as I saw it they only have functions for generating real numbers. So for the complex case I just generate double the number of real numbers.

**Review by hartwiganzt — APPROVED** [review]
As the remaining errors are potential division by zero, I think we can move this forward!

File: omp/solver/idr_kernels.cpp [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_IDR_STEP_2_KERNEL);
+
+
+template <typename ValueType>
+void step_3(std::shared_ptr<const OmpExecutor> exec, const size_type k,
+            const matrix::Dense<ValueType> *p, matrix::Dense<ValueType> *g,
+            matrix::Dense<ValueType> *g_k, matrix::Dense<ValueType> *u,
+            matrix::Dense<ValueType> *m, matrix::Dense<ValueType> *f,
+            matrix::Dense<ValueType> *alpha, matrix::Dense<ValueType> *residual,
```
Comment (tcojean) [inline omp/solver/idr_kernels.cpp:275]: Since you don't use alpha here, you can remove its name for clarity. Same for reference.

File: omp/test/solver/idr_kernels.cpp [hunk 18]
```diff
@@ -0,0 +1,421 @@
...
+}
+
+
+TEST_F(Idr, IdrStep3IsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::size_type k = 2;
+    gko::kernels::reference::idr::step_3(ref, k, p.get(), g.get(), v.get(),
+                                         u.get(), m.get(), f.get(), omega.get(),
```
Comment (tcojean) [inline omp/test/solver/idr_kernels.cpp:271]: In terms of naming, `omega` is `alpha` in the kernel parameter list?

File: omp/test/solver/idr_kernels.cpp [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+                                         r.get(), x.get(), stop_status.get());
+    gko::kernels::omp::idr::step_3(
+        omp, k, d_p.get(), d_g.get(), d_v.get(), d_u.get(), d_m.get(),
+        d_f.get(), d_omega.get(), d_r.get(), d_x.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(g, d_g, 1e-14);
+    GKO_ASSERT_MTX_NEAR(u, d_u, 1e-14);
```
Comment (tcojean) [inline omp/test/solver/idr_kernels.cpp:278]: Looks like `v` (`g_k` in the parameter list) is also computed on, so probably it should also be checked.
```suggestion
    GKO_ASSERT_MTX_NEAR(g, d_g, 1e-14);
    GKO_ASSERT_MTX_NEAR(v, d_v, 1e-14);
    GKO_ASSERT_MTX_NEAR(u, d_u, 1e-14);
```

File: cuda/test/solver/idr_kernels.cpp [hunk 20]
```diff
@@ -0,0 +1,416 @@
...
+    gko::size_type k = 2;
+    gko::kernels::reference::idr::step_3(ref, k, p.get(), g.get(), v.get(),
+                                         u.get(), m.get(), f.get(), omega.get(),
+                                         r.get(), x.get(), stop_status.get());
+    gko::kernels::cuda::idr::step_3(
+        cuda, k, d_p.get(), d_g.get(), d_v.get(), d_u.get(), d_m.get(),
+        d_f.get(), d_omega.get(), d_r.get(), d_x.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(g, d_g, 1e-14);
+    GKO_ASSERT_MTX_NEAR(u, d_u, 1e-14);
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:273]:
```suggestion
    GKO_ASSERT_MTX_NEAR(g, d_g, 1e-14);
    GKO_ASSERT_MTX_NEAR(v, d_v, 1e-14);
    GKO_ASSERT_MTX_NEAR(u, d_u, 1e-14);
```

File: hip/test/solver/idr_kernels.cpp [hunk 21]
```diff
@@ -0,0 +1,416 @@
...
+
+    GKO_ASSERT_MTX_NEAR(omega, d_omega, 1e-14);
+}
+
+
+TEST_F(Idr, IdrIterationOneRHSIsEquivalentToRef)
+{
+    int m = 123;
+    int n = 1;
+
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:301]:
```suggestion
```

File: hip/test/solver/idr_kernels.cpp [hunk 22]
```diff
@@ -0,0 +1,416 @@
...
+    GKO_ASSERT_MTX_NEAR(d_b, b, 1e-13);
+    GKO_ASSERT_MTX_NEAR(d_x, x, 1e-13);
+}
+
+
+TEST_F(Idr, IdrIterationWithComplexSubspaceOneRHSIsEquivalentToRef)
+{
+    int m = 123;
+    int n = 1;
+
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:323]:
```suggestion
```

File: hip/test/solver/idr_kernels.cpp [hunk 23]
```diff
@@ -0,0 +1,416 @@
...
+    hip_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(hip))
+            .on(hip);
+
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:331]:
```suggestion
```

File: hip/test/solver/idr_kernels.cpp [hunk 24]
```diff
@@ -0,0 +1,416 @@
...
+    ref_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(ref))
+            .on(ref);
+
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:339]:
```suggestion
```

File: hip/test/solver/idr_kernels.cpp [hunk 25]
```diff
@@ -0,0 +1,416 @@
...
+    GKO_ASSERT_MTX_NEAR(d_b, b, 1e-13);
+    GKO_ASSERT_MTX_NEAR(d_x, x, 1e-13);
+}
+
+
+TEST_F(Idr, IdrIterationWithComplexSubspaceMultipleRHSIsEquivalentToRef)
+{
+    int m = 123;
+    int n = 16;
+
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:382]:
```suggestion
```

File: hip/test/solver/idr_kernels.cpp [hunk 26]
```diff
@@ -0,0 +1,416 @@
...
+    hip_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(hip))
+            .on(hip);
+
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:390]:
```suggestion
```

File: hip/test/solver/idr_kernels.cpp [hunk 27]
```diff
@@ -0,0 +1,416 @@
...
+    ref_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(ref))
+            .on(ref);
+
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:398]:
```suggestion
```

File: hip/test/solver/idr_kernels.cpp [hunk 28]
```diff
@@ -0,0 +1,416 @@
...
+    gko::size_type k = 2;
+    gko::kernels::reference::idr::step_3(ref, k, p.get(), g.get(), v.get(),
+                                         u.get(), m.get(), f.get(), omega.get(),
+                                         r.get(), x.get(), stop_status.get());
+    gko::kernels::hip::idr::step_3(
+        hip, k, d_p.get(), d_g.get(), d_v.get(), d_u.get(), d_m.get(),
+        d_f.get(), d_omega.get(), d_r.get(), d_x.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(g, d_g, 1e-14);
+    GKO_ASSERT_MTX_NEAR(u, d_u, 1e-14);
```
Comment (tcojean) [inline hip/test/solver/idr_kernels.cpp:273]:
```suggestion
    GKO_ASSERT_MTX_NEAR(g, d_g, 1e-14);
    GKO_ASSERT_MTX_NEAR(v, d_v, 1e-14);
    GKO_ASSERT_MTX_NEAR(u, d_u, 1e-14);
```

File: cuda/test/solver/idr_kernels.cpp [hunk 29]
```diff
@@ -0,0 +1,416 @@
...
+
+    GKO_ASSERT_MTX_NEAR(omega, d_omega, 1e-14);
+}
+
+
+TEST_F(Idr, IdrIterationOneRHSIsEquivalentToRef)
+{
+    int m = 123;
+    int n = 1;
+
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:301]:
```suggestion
```

File: cuda/test/solver/idr_kernels.cpp [hunk 30]
```diff
@@ -0,0 +1,416 @@
...
+    GKO_ASSERT_MTX_NEAR(d_b, b, 1e-13);
+    GKO_ASSERT_MTX_NEAR(d_x, x, 1e-13);
+}
+
+
+TEST_F(Idr, IdrIterationWithComplexSubspaceOneRHSIsEquivalentToRef)
+{
+    int m = 123;
+    int n = 1;
+
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:323]:
```suggestion
```

File: cuda/test/solver/idr_kernels.cpp [hunk 31]
```diff
@@ -0,0 +1,416 @@
...
+    cuda_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(cuda))
+            .on(cuda);
+
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:331]:
```suggestion
```

File: cuda/test/solver/idr_kernels.cpp [hunk 32]
```diff
@@ -0,0 +1,416 @@
...
+    ref_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(ref))
+            .on(ref);
+
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:339]:
```suggestion
```

File: cuda/test/solver/idr_kernels.cpp [hunk 33]
```diff
@@ -0,0 +1,416 @@
...
+    GKO_ASSERT_MTX_NEAR(d_b, b, 1e-13);
+    GKO_ASSERT_MTX_NEAR(d_x, x, 1e-13);
+}
+
+
+TEST_F(Idr, IdrIterationWithComplexSubspaceMultipleRHSIsEquivalentToRef)
+{
+    int m = 123;
+    int n = 16;
+
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:382]:
```suggestion
```

File: cuda/test/solver/idr_kernels.cpp [hunk 34]
```diff
@@ -0,0 +1,416 @@
...
+    cuda_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(cuda))
+            .on(cuda);
+
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:390]:
```suggestion
```

File: cuda/test/solver/idr_kernels.cpp [hunk 35]
```diff
@@ -0,0 +1,416 @@
...
+    ref_idr_factory =
+        Solver::build()
+            .with_deterministic(true)
+            .with_complex_subspace(true)
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1u).on(ref))
+            .on(ref);
+
```
Comment (tcojean) [inline cuda/test/solver/idr_kernels.cpp:398]:
```suggestion
```

**Review by tcojean — APPROVED** [review]
LGTM! Some small comments.

I would be interested in profiling results of some kernels to see whether they are efficient, but I think we can focus on optimization in another PR (if needed).

File: core/solver/idr.cpp [hunk 36]
```diff
@@ -0,0 +1,287 @@
...
+        Vector::create(exec, gko::dim<2>(subspace_dim_, problem_size));
+
+    // Initialization
+    // m = identity
+    exec->run(idr::make_initialize(m.get(), subspace_vectors.get(),
+                                   deterministic_, &stop_status));
+
+    // omega = 1
+    exec->run(
+        idr::make_fill_array(omega->get_values(), nrhs, one<SubspaceType>()));
```
Comment (pratikvn) [inline core/solver/idr.cpp:151]: Can you not use the [`fill_array`](https://github.com/ginkgo-project/ginkgo/blob/e4b0b95a09b19de26a8af126acbb17344595e2bb/omp/components/fill_array.cpp#L43) ? Or rather why is this in the the `idr::` namespace ?

**Review by pratikvn — APPROVED** [review]
LGTM!

File: core/matrix/dense.cpp [hunk 37]
```diff
@@ -215,6 +215,39 @@ inline void conversion_helper(SparsityCsr<ValueType, IndexType> *result,
 
 }  // namespace
```
Comment (yhmtsai) [inline core/matrix/dense.cpp:217]:
```suggestion


```

File: core/solver/idr.cpp (hunk 36 above)
Comment (yhmtsai) [reply core/solver/idr.cpp:151]: fill_array is not in the public, so need to use it via `GKO_REGISTER_OPERATION`

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 38]
```diff
@@ -203,12 +210,18 @@ class RelativeResidualNorm : public ResidualNorm<ValueType> {
...
             NormVector::create_with_config_of(this->starting_tau_.get());
-        dense_rhs->compute_norm2(this->starting_tau_.get());
+        if (dynamic_cast<const matrix::Dense<to_complex<ValueType>> *>(
+                args.b.get())) {
+            auto dense_rhs = as<matrix::Dense<to_complex<ValueType>>>(args.b);
+            dense_rhs->compute_norm2(this->starting_tau_.get());
+        } else {
+            auto dense_rhs = as<Vector>(args.b);
+            dense_rhs->compute_norm2(this->starting_tau_.get());
+        }
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/residual_norm.hpp:224]: I am confused about these codes.
Could you explain the difference between if-else?

**Review by yhmtsai — CHANGES_REQUESTED** [review]
not yet finished

File: include/ginkgo/core/stop/residual_norm.hpp (hunk 38 above)
Comment (fritzgoebel) [reply include/ginkgo/core/stop/residual_norm.hpp:224]: When you use IDR with a complex subspace for a real ValueType, then `as<Vector>` expects a real vector for the `args.initial_residual`, but it is a complex vector. Also the system matrix has a different ValueType than the residual in this case, so I need to distinguish here.

Comment (yhmtsai) [reply include/ginkgo/core/stop/residual_norm.hpp:224]: I see. Thanks

File: reference/test/solver/idr_kernels.cpp [hunk 39]
```diff
@@ -0,0 +1,425 @@
...
+                              {-42.0, 57.0, 32.0, 89.0, 89.0, -39.0}},
+                             this->exec);
+    auto solver = this->idr_factory_precision->generate(locmtx);
+    auto b =
+        gko::initialize<Mtx>({0.0, -9.0, -2.0, 8.0, -5.0, -6.0}, this->exec);
+    auto x = gko::initialize<Mtx>({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, this->exec);
+
+    solver->apply(b.get(), x.get());
+
+
```
Comment (yhmtsai) [inline reference/test/solver/idr_kernels.cpp:259]:
```suggestion

```

File: reference/test/solver/idr_kernels.cpp [hunk 40]
```diff
@@ -0,0 +1,425 @@
...
+    auto one_op = gko::initialize<gko::matrix::Dense<value_type>>(
+        {gko::one<value_type>()}, this->exec);
+    auto neg_one_op = gko::initialize<gko::matrix::Dense<value_type>>(
+        {-gko::one<value_type>()}, this->exec);
+    auto resnorm = gko::matrix::Dense<gko::remove_complex<value_type>>::create(
+        this->exec, gko::dim<2>{1, 1});
+    locmtx->apply(neg_one_op.get(), x.get(), one_op.get(), b.get());
+
```
Comment (yhmtsai) [inline reference/test/solver/idr_kernels.cpp:267]: I am not sure whether keep this line in AAA sense.

File: common/solver/idr_kernels.hpp.inc [hunk 41]
```diff
@@ -0,0 +1,348 @@
...
+{
+    const auto tidx = thread::get_thread_id_flat();
+
+    __shared__ UninitializedArray<ValueType, block_size> reduction_helper_array;
+    ValueType *__restrict__ reduction_helper = reduction_helper_array;
+
+    __shared__ UninitializedArray<remove_complex<ValueType>, block_size>
+        reduction_helper_array_real;
+    remove_complex<ValueType> *__restrict__ reduction_helper_real =
+        reduction_helper_array_real;
```
Comment (yhmtsai) [inline common/solver/idr_kernels.hpp.inc:67]: It is remove_complex type, so you can directly use `__shared__ remove_complex<ValueType> ...`

File: cuda/solver/idr_kernels.cu (hunk 9 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply cuda/solver/idr_kernels.cu:133]: I see

**Review by yhmtsai — APPROVED** [review]
just leave a reminder here: we can leave the subspace as user input.
They can generate from different seed or generator, etc.
We can add it afterwards if it is needed.
The PR looks good to me.
